### PR TITLE
[GSOC] Oauth2 google idp

### PR DESCRIPTION
**Pull request by neogismm**
### Description

This PR is to add functionality for OAuth 2.0 authentication with Google as IDP.

Progress:

- [x] Obtain client id/secret from Google API console
- [x] Create a Util class for the core methods
- [x] Create a Servlet class that handles redirect URLs for a successful or failed auth flow
- [x] Method to check if the user is already logged in
- [ ] Check if the user is a new or existing user (implement checking methods with cloudstack database)
- [x] Logout method
- [ ] Make changes in UI

NB: Progress milestones to be added/edited as I grow my understanding of the codebase.
 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [x] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

File: plugins/user-authenticators/oauth2/src/main/java/org/apache/cloudstack/oauth2/OAuth2Utils.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import com.google.api.client.http.HttpTransport;
+import com.google.api.client.http.javanet.NetHttpTransport;
+import com.google.api.client.json.jackson2.JacksonFactory;
+import com.google.api.client.util.store.MemoryDataStoreFactory;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+
+public class OAuth2Utils {
```
Comment (nvazquez) [inline plugins/user-authenticators/oauth2/src/main/java/org/apache/cloudstack/oauth2/OAuth2Utils.java:13]: This class looks very specific to Google, we can rather rename it or let the Google auth be the implementation of an abstract method in the oauth framework

Comment (neogismm) [reply plugins/user-authenticators/oauth2/src/main/java/org/apache/cloudstack/oauth2/OAuth2Utils.java:34]: Yes we can do that. AuthorizationCodeFlow class can be used to connect to any oauth server. So it makes sense to keep only the generic methods in the util class. But as of now, I'm keen towards building a minimum viable product with only google as idp. Once it is done, I'll tidy up the Util class to only include methods that can be used by any oauth server.

File: engine/schema/src/main/resources/META-INF/db/schema-41700to41710.sql [hunk 2]
```diff
@@ -20,4 +20,112 @@
 --;
 
 UPDATE `cloud`.`configuration` SET `value` = 'false' WHERE `name` = 'network.disable.rpfilter' AND `value` != 'true';
-UPDATE `cloud`.`configuration` SET `value` = 'false' WHERE `name` = 'consoleproxy.disable.rpfilter' AND `value` != 'true';
\ No newline at end of file
+UPDATE `cloud`.`configuration` SET `value` = 'false' WHERE `name` = 'consoleproxy.disable.rpfilter' AND `value` != 'true';
+
```
Comment (DaanHoogland) [inline engine/schema/src/main/resources/META-INF/db/schema-41700to41710.sql:24]: why changes for the DB v 4.17.1 in this PR?

Comment (neogismm) [reply engine/schema/src/main/resources/META-INF/db/schema-41700to41710.sql:24]: Rebasing the branch and pushing brings changes from all authors including my own changes.
My changes are reserved for plugins/user-authenticators/oauth2.
I know this is not the right way to push my code, can you suggest how to do it the right way?
Maybe I'll create a new PR with only my changes and close this one.

Comment (DaanHoogland) [reply engine/schema/src/main/resources/META-INF/db/schema-41700to41710.sql:24]: sorry @neogismm I do not understand the entire background. Are you saying you included code for other branches in this one? If you base your changes on main that shouldn´t happen. You can clean up your local working branch and force push it to your github branch to clean it up. There is no need for a new PR. 
But these schema changes for 4.17 are not wanted, i'd say.

Comment (neogismm) [reply engine/schema/src/main/resources/META-INF/db/schema-41700to41710.sql:24]: Sorry @DaanHoogland but I'm not able to figure out how to clean up this branch so I'm closing this PR and opening a new one that only includes my additions/changes to the OAuth feature.

**Pull request closed without merging**